source/mgnp-pitch/tests/support.rs

```diff
@@ -0,0 +1,245 @@
+                    let InboundConnect { hello, rsp } = req;
+                    tracing::info!(?hello, "hello world service received connection");
+                    let (their_chan, my_chan) =
+                        make_bidis::<HelloWorldRequest, HelloWorldResponse>(8);
+                    tokio::spawn(
+                        async move {
+                            tracing::debug!("hello world worker running...");
+                            while let Some(req) = my_chan.rx().recv().await {
+                                tracing::info!(?req);
+                                assert_eq!(req.hello, "hello");
+                                my_chan
+                                    .tx()
+                                    .send(HelloWorldResponse {
+                                        world: "world".into(),
+                                    })
+                                    .await
+                                    .unwrap();
+                            }
+                        }
+                        .instrument(tracing::info_span!("hello_world_worker", worker)),
+                    );
+                    worker += 1;
+                    let sent = rsp.send(Ok(their_chan)).is_ok();
+                    tracing::debug!(?sent);
+                }
+            }
+            .instrument(tracing::info_span!("hello_world_service")),
+        );
+    }
+}
+
+// === impl TestWire ===
+
+impl TestWire {
+    pub fn new() -> (Self, Self) {
+        let (tx1, rx1) = mpsc::channel(8);
+        let (tx2, rx2) = mpsc::channel(8);
+        (Self { tx: tx1, rx: rx2 }, Self { tx: tx2, rx: rx1 })
+    }
+}
+
+impl Wire for TestWire {
+    type Frame = TestFrame;
+    type Error = &'static str;
+
+    async fn recv(&mut self) -> Result<Self::Frame, &'static str> {
+        let frame = self.rx.recv().await;
+        tracing::info!(frame = ?frame.as_ref().map(HexSlice::new), "RECV");
+        frame
+            .ok_or("the send end of this wire has been dropped")
+            .map(TestFrame)
+    }
+
+    async fn send(&mut self, msg: OutboundMessage<'_>) -> Result<(), &'static str> {
+        // TODO(eliza): this is awkward, the trickypipe SerRecvRef API needs to
+        // suck less so we don't need to do it like this...
+        tracing::info!(?msg, "sending message");
+        let frame = match msg {
+            OutboundMessage::Control(ctrl) => {
+                postcard::to_allocvec(&InboundMessage::Control(ctrl)).unwrap()
+            }
+            OutboundMessage::Data {
+                local_id,
+                remote_id,
+                data,
+            } => {
+                let data = data.to_vec().unwrap();
+                postcard::to_allocvec(&InboundMessage::Data {
+                    local_id,
+                    remote_id,
+                    data: &data[..],
+                })
+                .unwrap()
+            }
+        };
+
```

Comment:
the fact that we have to serialize into a vec here rather than writing directly to the "wire" (which might be e.g. a byte pipe) is a huge deficiency of the current impl. we need to fix this before we're able to use this stuff in "real life" --- the outbound message should be able to be serialized to a byte slice or directly to an `io::Write`/`embedded_io::Write`/etc...

Reply:
I *might* have a hack/an idea for this, in one of the branches where I was trying to get no-alloc stuff working. Let me look at it with fresh eyes tomorrow.